### PR TITLE
Use NewFilePath instead of NewFile while creating the File object

### PR DESCRIPTION
**Pull request by sergenyalcin**
Signed-off-by: Sergen Yalçın <yalcinsergen97@gmail.com>

### Description of your changes

Fixes https://github.com/crossplane/terrajet/issues/228

While writing the files, the `NewFile` function was used for creating the `File` object. This function takes only the `packageName` as parameter. When we create the `File` object this function, since the tool does not have the entire file path, it cannot detect whether the import is from the same package.

In this PR, `NewFilePath` function is used instead of `NewFile` while creating the `File` object. `NewFilePath` function takes the `packagePath` as parameter. By this way, while determining the imports, the tool can detect whether the import path is from the same package.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

The two test cases were applied for this PR:
- Generate the provider-jet-aws resources again to check there is not any difference.
- `aws_iam_access_key` resource has `aws_iam_user` reference:

```
r.References = config.References{
	"user": config.Reference{
		Type: "User",
	},
}
```

This block was changed with the following one:

```
r.References = config.References{
	"user": config.Reference{
		Type: "github.com/crossplane-contrib/provider-jet-aws/apis/iam/v1alpha2.User",
	},
}
```

In new configuration the full path of the reference package was used, then the resources were generated. This case is also successful.

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Comment by sergenyalcin** [comment]
@muvaf as we discussed offline, the our dependent tool can handle the same package imports. However, to use this feature we need to create File object by passing the entire package path. So `NewFilePath` function was used in this PR.

File: internal/generate/generate.go [hunk 1]
```diff
@@ -83,7 +83,7 @@ func WriteMethods(p *packages.Package, ms method.Set, file string, wo ...WriteOp
 		fn(opts)
 	}
 
-	f := jen.NewFile(p.Name)
+	f := jen.NewFilePath(p.PkgPath)
```
Comment (ulucinar) [inline internal/generate/generate.go:86]: Could you please explain with a comment the behavior difference between `jen.NewFile` and `jen.NewFilePath`? My understanding is that we need `NewFilePath` so that we can communicate the package to the library (`jennifer`) not just the package name.

File: internal/method/resolver_test.go [hunk 2]
```diff
@@ -220,6 +226,22 @@ func (mg *Model) ResolveReferences(ctx context.Context, c client.Reader) error {
 	mg.Spec.ForProvider.RouteTableIDs = reference.ToPtrValues(mrsp.ResolvedValues)
 	mg.Spec.ForProvider.RouteTableIDsRefs = mrsp.ResolvedReferences
 
+	rsp, err = r.Resolve(ctx, reference.ResolutionRequest{
```
Comment (ulucinar) [inline internal/method/resolver_test.go:229]: Could you please consider adding tests for the `generate.WriteMethods` function?

File: internal/method/resolver_test.go [hunk 3]
```diff
@@ -238,7 +260,7 @@ func TestNewResolveReferences(t *testing.T) {
 	if err != nil {
 		t.Error(err)
 	}
-	f := jen.NewFile("v1alpha1")
+	f := jen.NewFilePath("golang.org/fake/v1alpha1")
```
Comment (ulucinar) [inline internal/method/resolver_test.go:263]: It looks like that something is wrong in theory here. We are fixing the `generate.WriteMethods` method but we need to also make the same change in the test's own implementation. Could you please explain what's going on here? This test case does not look like to be related to the `generate.WriteMethods` change.

Comment (sergenyalcin) [reply internal/method/resolver_test.go:263]: As you mentioned, this test does not test the `WriteMethods` function. In other words, the tests are not affected due to the change made in the production code (`WriteMethods` function). And if changes made from this test code are rolled back, the tests will pass. Because this test has nothing to do with `WriteMethods`.

So why was this change made in this test? As you know, in the issue that requires this fix, the import cycle problem occurs when the entire reference path for the resource configuration is passed and this reference is in the same package as the related resource. So I added a new case to the unit test of the function (in `NewResolveReferences`) that resolves the References and generates their functions, regardless of the `WriteMethods` change that is the subject of the PR. This case contains a referenced case in the same package:

```
// +crossplane:generate:reference:type=golang.org/fake/v1alpha1.Configuration
// +crossplane:generate:reference:extractor=golang.org/fake/v1alpha1.Configurations()
Configurations *Configuration
```

After the mentioned new case was added to the unit test, the test started to fail. Therefore, a change in testing was required.

In summary, as you said, there is no need to change this test specifically for the fix of this issue. Perhaps it would make sense to remove the change from the test file to avoid confusion.

It seems like it would make more sense to make these changes in a separate PR that deals with the refactoring of the tests.

File: internal/generate/generate.go (hunk 1 above)
Comment (sergenyalcin) [reply internal/generate/generate.go:93]: Done!

File: internal/method/resolver_test.go (hunk 2 above)
Comment (sergenyalcin) [reply internal/method/resolver_test.go:229]: In `WriteMethods` there is a file writing operation: 

`return errors.Wrap(ioutil.WriteFile(file, b.Bytes(), 0644), "cannot write Go file")`

It seems that when we want to write unit-tests to this function we will have to find a good way to handle this operation (such as mocking this operation or writing test files with temporary way).

It looks like you need to make changes to the production code to use a mock method. Because currently, the function accepts a filename and directly calls ioutil's Write function.

I suggest that this unit-test be followed by opening another issue together with the refactoring in the prod code. What are your thoughts @ulucinar

File: internal/method/resolver_test.go (hunk 3 above)
Comment (muvaf) [reply internal/method/resolver_test.go]: I think if we use `NewFilePath` in the production path but `NewFile` in the unit tests, then it'd invalidate the test's assumptions since we're trying to emulate the prod flow in the test file so that it doesn't end up being an integration test for the whole thing, so changing all tests to use `NewFilePath` seems like the right thing to do to me.

Comment (sergenyalcin) [reply internal/method/resolver_test.go:261]: @muvaf I changed the all tests to use `NewFilePath`. Also for `TestNewResolveReferences` added a new case about same package reference.

**Review by muvaf — APPROVED** [review]
LGTM! Thanks @sergenyalcin !
